Add docs for MessagePack (#21090)
dotnet/AspNetCore.Docs: aspnetcore/signalr/java-client.md

When calling one of these methods with one or more object types, use the generics syntax when invoking the method. For example, when registering an `on` handler for a method named `func`, which takes as arguments a String and a `Foo<String>` object, use the following code to set an action to print the arguments:

```java
hubConnection.<String, Foo<String>>on("func", (param1, param2) ->{
    System.out.println(param1);
    System.out.println(param2);
}, String.class, fooType);
```

This convention is necessary because we can not retrieve complete information about complex types with the `Object.getClass` method due to type erasure in Java. For example, calling `getClass` on an `ArrayList<String>` would not return `Class<ArrayList<String>>`, but rather `Class<ArrayList>`, which does not give the deserializer enough information to correctly deserialize an incoming message. The same is true for custom objects.

::: moniker-end

## Known limitations

::: moniker range=">= aspnetcore-5.0"

* Transport fallback and the Server Sent Events transport aren't supported.

::: moniker-end

> ::: moniker range=">= aspnetcore-3.0" < aspnetcore-5.0"

Author: @Rick-Anderson the build is complaining about invalid characters here, did I do something wrong with the moniker sections?

Reviewer: ```suggestion
::: moniker range=">= aspnetcore-3.0 < aspnetcore-5.0"
```

Reviewer 2: There's an extra double quote after "3.0". Removing that should fix the error.

    })).build();
```

::: moniker range=">= aspnetcore-5.0"

### Passing Class information in Java

When calling the `on`, `invoke`, or `stream` methods of `HubConnection` in the Java client, users should pass a `Type` object rather than a `Class<?>` object to describe each `Object` passed to the method. A `Type` can be acquired using the provided `TypeReference` class. For example, using a custom classes named `Foo` and `Bar`, the following code gets each `Type`:

```java
Type fooType = new TypeReference<Foo>() { }).getType();
Type barType = new TypeReference<Bar>() { }).getType();

> ```

Reviewer: This only necessary for generic types, right? I think `Bar.class` would work fine.

Author: Oops, yes, you're right

When calling one of these methods with one or more object types, use the generics syntax when invoking the method. For example, when registering an `on` handler for a method named `func`, which takes as arguments a `Foo` object and a `Bar` object, use the following code to set an action to print the two objects:

```java
hubConnection.<Foo, Bar>on("func", (param1, param2) ->{
    System.out.println(param1);
    System.out.println(param2);
}, fooType, barType);
```

> The reason that this convention is necessary is, due to type erasure in Java, we can not retrieve complete information about complex types with the `Object.getClass` method. For example, calling `getClass` on an `ArrayList<String>` would not return `Class<ArrayList<String>>`, but rather `Class<ArrayList>`, which does not give the deserializer enough information to correctly deserialize an incoming message. The same is true for custom objects.

Reviewer: Nit:

```suggestion
This convention is necessary because we can not retrieve complete information about complex types with the `Object.getClass` method due to type erasure in Java. For example, calling `getClass` on an `ArrayList<String>` would not return `Class<ArrayList<String>>`, but rather `Class<ArrayList>`, which does not give the deserializer enough information to correctly deserialize an incoming message. The same is true for custom objects.
```

::: moniker range=">= aspnetcore-3.0"

> * Only the JSON protocol is supported.

Reviewer 3: Should add a 5.0 moniker and remove it there and keep it for 3.1
